python/hail/docs/tutorials/08-joins.ipynb

```diff
@@ -0,0 +1,377 @@
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "## Keys\n",
+    "\n",
+    "To understand joins in Hail, we need to revisit one of the crucial properties of `Table`s: the key.\n",
+    "\n",
+    "A `Table` has an ordered list of fields known as the key.  The key is show by `describe`."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "users.describe()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "`key` is a struct expression of all of the key fields."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "users.key.describe()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "Hail makes no assumptions on keys.  In particular, the keys need not be unique or non-missing, although in many applications they will be both.\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "To understand Hail's join syntax, it is easiest to see an example."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "t1 = hl.Table.parallelize([\n",
+    "    {'a': 'foo', 'b': 1},\n",
+    "    {'a': 'bar', 'b': 2},\n",
+    "    {'a': 'bar', 'b': 2}],\n",
+    "    hl.tstruct(a=hl.tstr, b=hl.tint32),\n",
+    "    key='a')\n",
+    "t2 = hl.Table.parallelize([\n",
+    "    {'t': 'foo', 'x': 3.14},\n",
+    "    {'t': 'bar', 'x': 2.78},\n",
+    "    {'t': 'bar', 'x': -1},\n",
+    "    {'t': 'quam', 'x': 0}],\n",
+    "    hl.tstruct(t=hl.tstr, x=hl.tfloat64),\n",
+    "    key='t')"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "t1.show()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "t2.show()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "j = t1.annotate(t2_x = t2[t1.a].x)\n",
+    "j.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "The magic of keys is that they turn tables into maps: `table[expr]` should naturally refer to the row of `table` that has key the value of `expr`.  Note: if the row is not unique, one such row is chosen arbitrarily.\n",
```

Comment:
Would be good to note that despite the notational similarity, this is different than the `table['field']` notation with a string.